### ISSUE TITLE
questions about instance normalization in Generator of CycleGAN

### ISSUE THREAD

**Issue by mohammadhosseinashoori** [junyanz/pytorch-CycleGAN-and-pix2pix#981]
hello, i have two questions:
**1)** I understand that why you ignore the InstanceNorm in the first block of Unet (Pix2Pix)

> If you use an instancenorm in the first layer, the color of the input image will be normalized and get ignored. For many applications, you may want to preserve the color of the input image

junyanz said at : https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/issues/782

**okay, but why you apply InstanceNorm in the first block of Resnet (CycleGAN) ???**

**2)** use_bias = norm_layer == nn.InstanceNorm2d
I dont understand this line of code ...
I think use_bias of conv layer must be False before batchnorm layer or instancenorm (both in affine = true or false)

can you explain it?

Thanks a lot in Advance

**Comment by junyanz** [comment]
1) For CycleGAN, we followed the resnet-based architecture from prior [work](https://arxiv.org/abs/1603.08155). It has a large Conv layer (7x7) before the norm layer, which may be able to encode color information. Feel free to remove this layer and see how it works. @taesungp may know more details. 
2) It depends on the flag `affine`. If `affine=True`, the normalization layer already includes bias parameters and we do not need `use_bias` for the Conv layers before that normalization layer.  By default, we set `affine=False` for `nn.InstancNorm2d`.

**Comment by mohammadhosseinashoori** [comment]
thank you for your replay.

**1) ok**
**2)** I think, after normalization, bias term will be ignored, wheater affine = True or False

in SPADE project, they removed bias before instance norm (used in discriminator):

https://github.com/NVlabs/SPADE/blob/master/models/networks/normalization.py 
(Line 33, at get_nonspade_norm_layer function)

>         # remove bias in the previous layer, which is meaningless
>         # since it has no effect after normalization
>         if getattr(layer, 'bias', None) is not None:
>             delattr(layer, 'bias')
>             layer.register_parameter('bias', None)

i get confused, ...
thank you

**Comment by junyanz** [comment]
@taesungp might be able to answer your question regarding SPADE.

**Comment by taesungp** [comment]
Hi @mohammadhosseinashoori, I think you are correct in that the bias of the conv layer right before IN is not needed, as it will be canceled in the normalization layer. However, in CycleGAN, we went with the default setting of nn.Conv, which enables bias by default. It was just for simplicity of the coding. Later on we integrated a PR that removes the bias, but the bias for IN was likely still kept to be able to run pretrained models.
